### PR TITLE
More click modes (with tests!)

### PR DESCRIPTION
**Pull request by kaduvert**
Hey there,

this pull request implements:

shift clicks,
number clicks,
middle clicks
and drop clicks

or mode 1-4 as seen in

this image

with typescript definitions, **tests (mocha)** and improved code readability.

You may recall me from a pull request one and a half years ago, which closed due to lack of tests and maybe other stuff.
I finally got around to working on it and i think i've addressed all the issues! :)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ function loader (mcVersion) {
   }
 
   const windowByType = new Map()
-  for (const key of Object.keys(windows)) {
+  for (const key in windows) {
```
Comment (rom1504) [inline index.js]: don't use in, in javascript it's bad
prefer for or .forEach

**Comment by rom1504** [comment]
looks ok

need more review

could you check @u9g maybe ?

File: index.js (hunk 1 above)
Comment (kaduvert) [reply index.js]: sorry for that one, removed.

File: index.d.ts [hunk 2]
```diff
@@ -132,6 +218,13 @@ export class Window<T = unknown> extends (EventEmitter as new <T>() => TypedEmit
      */
     firstEmptyInventorySlot(hotbarFirst?: boolean): number | null;
 
+    /**
+     * Returns how much items there are ignoring what the item is
+     * @param start
+     * @param end
+     */
+    sumRange(start: number, end: number): number;
```
Comment (u9g) [inline index.d.ts:254]: can you give a bit more info on what this is counting? what are start and end?

File: index.js [hunk 3]
```diff
@@ -67,8 +67,7 @@ function loader (mcVersion) {
 
   return {
     createWindow: (id, type, title, slotCount = undefined) => {
-      let winData = windowByType.get(type)
-      if (!winData) winData = windows[type]
+      let winData = windowByType.get(type) || windows[type]
```
Comment (u9g) [inline index.js]:
```suggestion
      let winData = windowByType.get(type) ?? windows[type]
```

File: lib/Window.js [hunk 4]
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
     }
 
     acceptClick (click) {
-      assert.ok(click.mouseButton === 0 || click.mouseButton === 1)
+      const { mode, slot, mouseButton } = click
+      assert.ok(
+        (mode <= 6 && mode >= 0) &&
```
Comment (u9g) [inline lib/Window.js]:
```suggestion
        (mode >= 0 && mode <= 6) &&
```

File: lib/Window.js [hunk 5]
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
...
+          this.mouseClick(click)
+          break
+
+        case 1:
+          assert.ok(mouseButton <= 1, 'invalid operation')
+          this.shiftClick(click)
+          break
+
+        case 2:
+          assert.ok(mouseButton <= 8, 'invalid operation')
```
Comment (u9g) [inline lib/Window.js:50]: You're ignoring mode 2, button 40, which is offhand swap key

File: lib/Window.js [hunk 6]
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
...
+          this.middleClick(click)
+          break
+
+        case 4:
+          assert.ok(mouseButton <= 1, 'invalid operation')
+          this.dropClick(click)
+          break
+
+        case 5:
+          this.dragClick(click)
```
Comment (u9g) [inline lib/Window.js]: if you're doing asserts on the rest, might as well assert that this is in [1,5,9,2,6,10]

File: lib/Window.js [hunk 7]
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
...
       if (click.slot === -999) {
-        return this.acceptOutsideWindowClick(click)
-      } else if (click.slot >= this.inventoryStart && click.slot < this.inventoryEnd) {
-        return this.acceptInventoryClick(click)
-      } else if (click.slot === this.craftingResultSlot) {
-        return this.acceptCraftingClick(click)
+        this.dropSelectedItem(click.mouseButton === 0)
       } else {
-        return this.acceptNonInventorySwapAreaClick(click)
+        let { item } = click
```
Comment (u9g) [inline lib/Window.js:83]: why were the smaller methods removed?

**Review by u9g — CHANGES_REQUESTED** [review]
Some thoughts, I'll finish the review after these get fixed.

File: index.d.ts (hunk 2 above)
Comment (kaduvert) [reply index.d.ts:254]: Returns how much items there are ignoring what the item is

basically taking start slot and end slot and counting the items from start to end slot, but i think just reading the function code would make this obvious

edit: i clarified that in the typescript definitions a bit

File: lib/Window.js (hunk 7 above)
Comment (kaduvert) [reply lib/Window.js:83]: they were hard to read and inflexible.

i read the entire old code and figured a complete rewrite of the functions directly handling clicks is the best option.

File: lib/Window.js (hunk 5 above)
Comment (kaduvert) [reply lib/Window.js:50]: in minecraft 1.8 there is no offhand,
i just implemented clicks for 1.8, as i have no experience with minecraft above version 1.8 and don't feel confident implementing that

**Comment by KadaverBrutalo10** [comment]
@u9g

File: index.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
 function loader (registryOrVersion) {
   const registry = typeof registryOrVersion === 'string' ? require('prismarine-registry')(registryOrVersion) : registryOrVersion
-  const Item = require('prismarine-item')(registry)
```
Comment (IceTank) [inline index.js:3]: Why was this changed? It is more efficient if prismarine-item can use the same registry version as p window.

File: lib/Window.js [hunk 9]
```diff
@@ -23,105 +23,205 @@ module.exports = (Item) => {
...
+              this.splitSlot(item)
+            } else {
+              this.swapSelectedItem(click.slot, item)
+            }
+          }
+        }
       }
     }
 
-    acceptOutsideWindowClick (click) {
```
Comment (IceTank) [inline lib/Window.js:38]: Changing all the already implemented function to new functions with different names is a breaking change. We won't be able to release this without a major version jump or this will break everything that is using prismarine-windows.

**Comment by IceTank** [comment]
Are new function implementation really necessary? If prismarine-windows is changing everything we also have to re implement windows in mineflayer.

File: index.js (hunk 8 above)
Comment (kaduvert) [reply index.js:3]: because of https://github.com/PrismarineJS/prismarine-windows/pull/77

i tested my pr with these changes and my tests failed,
i quickly looked up how to get the mcVersion and implemented that so everything worked again.

i don't know anything about this refactoring i just made my pr compatible with it
if you want you can edit it, but make sure the tests all succeed

Comment (IceTank) [reply index.js:3]: I will look at it. But I don't see a reason why that would fail. Maybe you where not using up to date dependencies?

File: lib/Window.js (hunk 9 above)
Comment (kaduvert) [reply lib/Window.js:38]: As stated above "they were hard to read and inflexible"

The functions that were removed should only be called by acceptClick() and are probably also not called from anywhere outside, as the removed functions simply split up the logic of acceptClick()

**Comment by kaduvert** [comment]
@IceTank Yes, i think so.
The old functions were unworkable with,
that's the reason i removed them and likely also the reason why nobody else implemented other click types for a long time

Also this doesn't break mineflayer, if mineflayer was to use this then it would work as before, because the only clicks currently allowed by mineflayer are normal ones (mode 0) and they work as they did before

File: index.js (hunk 8 above)
Comment (kaduvert) [reply index.js:3]: i think it was because prismarine-item wanted a mcversion string instead of some registry object

**Comment by rom1504** [comment]
@u9g @IceTank @KadaverBrutalo10 thoughts: ready to merge ?

**Comment by kaduvert** [comment]
@rom1504 sure

**Comment by IceTank** [comment]
Still looks like this is a breaking change in the API. There might be some projects that break even with the small API changes.
This does looks like a good change tho and we should definitely add more click options.
But, as I understand, this was only made and tested for version 1.8. We need to test how well it works for newer versions too. 
Maybe someone can do a bigger manual test? We could just jump head first into it and try to patch stuff after but I have a feeling that this could break a lot of things in mineflayer.

**Comment by KadaverBrutalo10** [comment]
Well you see, for projects to break, projects first have to depend on the removed functionality in the first place.
Just to make sure we're talking about the same thing, these are the functions that would be removed:

```js
acceptOutsideWindowClick(click)
acceptInventoryClick(click)
acceptNonInventorySwapAreaClick(click)
acceptSwapAreaRightClick(click)
acceptSwapAreaLeftClick(click)
```

These are all functions, that are called within prismarine-windows, and within prismarine-windows only.
The purpose of `acceptClick(click)` is, to split up internal click-handling logic.
Code outside of prismarine-windows **should not call** these functions.

The only function it makes sense to call from outside is `acceptClick(click)` to accept a click of any kind.

So if any projects depend on these functions:
first of all, they shouldn't, but
second of all, they can just call `acceptClick(click)` instead! (they should be doing this already)
Currently, **not even mineflayer (v4.8.1)** calls *any* of the removed functions directly.

As to the testing side of things, the currently supported click modes (just 0) **work as they did before**, because the logic remains unchanged. Sure, the function names are different and all that (for improved read- and reuse-ability), but the raw logic remains the same.

Because things are taking long (~1.5 yrs) i did extensive testing of my changes already through normal use.
Over time, others even *starred and forked* my fork, so it doesn't seem to be so bad after all...

Why don't you go ahead and test/review my changes? Why aren't you the one doing the "bigger manual tests"?

**Comment by KadaverBrutalo10** [comment]
sorry @IceTank if my last comment sounded a bit aggressive. :(
i likely even commented on statements, that weren't even directed at me.
i just wanted to make it clear, that my changes are not causing major changes to mineflayer and also not to any realistic projects using prismarine-windows.

to make up for that (and maybe also because i need it), i implemented middleClicks :)

**Comment by kaduvert** [comment]
forget what i said about how unreasonable it would be to call the removed functions, i added them back in.

they now are all aliases of acceptClick... and yes, this is perfectly fine.
hope everyone's happy now and merging can begin! :)

File: test/test.js [hunk 10]
```diff
@@ -0,0 +1,141 @@
+const mocha = require('mocha')
+const { describe, it } = mocha
```
Comment (IceTank) [inline test/test.js]:
```suggestion
const { describe, it } = require('mocha')
```

File: test/test.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+const assert = require('assert')
+
+const windows = require('..')('1.8')
```
Comment (IceTank) [inline test/test.js]: Have you tested this with other versions as well?

File: test/test.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+const createClick = (mode, mouseButton, slot, useInv = false) => ({
+  slot,
+  mouseButton,
+  mode,
+  windowId: useInv ? inv.id : chest.id,
+  item: slot === -999 ? null : useInv ? inv.slots[slot] : chest.slots[slot]
+})
+
+chest.updateSlot(0, new Item(1, 64))
```
Comment (IceTank) [inline test/test.js]: Please don't use hard coded item id's. I know this is a 1.8 specific test but item id's change between versions. It is better to use `mcData.itemsByName.dirt.id` for instance.

File: test/test.js [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+      assert(!chest.slots[0] && chest.slots[1]?.count === 64 && chest.slots[2]?.count === 3)
+    })
+
+    it('shift out of inventory into chest #2', () => {
+      chest.acceptClick(createClick(1, 0, chest.inventoryEnd - 2))
+      assert(!chest.slots[chest.inventoryEnd - 2] && chest.slots[2]?.count === 64 && chest.slots[0]?.count === 3)
+    })
+
+    it.skip('shift out of inventory into armor slot - leather boots', () => {
+      // isn't even implemented, so it would naturally fail.
```
Comment (IceTank) [inline test/test.js]: Is the implementation planed? Or is there an active issue for this?

File: test/test.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+chest.updateSlot(0, new Item(1, 64))
+chest.updateSlot(1, new Item(1, 1))
+chest.updateSlot(3, new Item(1, 63))
+chest.updateSlot(5, new Item(2, 3))
+chest.updateSlot(chest.inventoryEnd - 1, new Item(1, 44))
+chest.updateSlot(chest.inventoryEnd - 2, new Item(1, 44))
+chest.updateSlot(chest.inventoryEnd - 3, new Item(1, 44))
+
+inv.updateSlot(inv.inventoryEnd - 1, new Item(301, 1))
+inv.updateSlot(inv.inventoryEnd - 2, new Item(1, 1))
```
Comment (IceTank) [inline test/test.js]: So the chest content and inventory content is initialized as the test is loaded. What happens when clicks are accepted? Do the items move inside of the container? Maybe it is better to use mocha's `beforeEach()` hook? https://mochajs.org/#hooks

You can run specific tests with mocha by using regex to match the test name. Do the tests break if they are not all executed after each other?

File: package.json [hunk 15]
```diff
@@ -4,7 +4,7 @@
   "description": "Represent minecraft windows",
   "main": "index.js",
   "scripts": {
-    "test": "npm run lint",
+    "test": "npm run lint && mocha",
```
Comment (IceTank) [inline package.json]: Test's in prismarine projects are usually implemented with scripts that look like this 
```json
"mocha_test": "mocha --reporter spec --exit",
"test": "npm run mocha_test",
```
See mineflayers package.json file https://github.com/PrismarineJS/mineflayer/blob/master/package.json

**Review by IceTank — CHANGES_REQUESTED** [review]

File: test/test.js (hunk 11 above)
Comment (kaduvert) [reply test/test.js]: no

to my knowledge, the click modes implemented here have not changed much since 1.8

File: test/test.js (hunk 13 above)
Comment (kaduvert) [reply test/test.js]: neither,
there is no issue because in the current prismarine-windows you'd have to explicitly trigger this behaviour.
as to implementing, i couldn't really figure out how to do it, as i didn't find any mcData stuff which describes which items can go into which slot

an issue should probably be created for this.

File: test/test.js (hunk 14 above)
Comment (kaduvert) [reply test/test.js]: yes, the tests absolutely suck and will break if they are not executed in order xD

i have no idea how mocha works, just wanted to make minimal tests, which show that my changes work.
if you understand the logic behind my tests, you are likely able to improve them by a lot

File: test/test.js (hunk 10 above)
Comment (kaduvert) [reply test/test.js]: implemented.

File: test/test.js (hunk 12 above)
Comment (kaduvert) [reply test/test.js]: implemented.

File: package.json (hunk 15 above)
Comment (kaduvert) [reply package.json]: implemented.

File: test/test.js [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+chest.updateSlot(chest.inventoryEnd - 3, new Item(stoneId, 44))
+
+inv.updateSlot(inv.inventoryEnd - 1, new Item(registry.itemsByName.leather_boots.id, 1))
+inv.updateSlot(inv.inventoryEnd - 2, new Item(stoneId, 1))
+
+describe('mode 0 | normal click', () => {
+  describe('mouseButton 0', () => {
+    it('pickup item', () => {
+      chest.acceptClick(createClick(0, 0, 0))
+      assert(!chest.slots[0] && chest.selectedItem?.count === 64 && chest.selectedItem.type === 1)
```
Comment (IceTank) [inline test/test.js]: Don't put multiple assertions into one assert. This will make a test unreadable if it does fail. The assertion is also missing a meaningful comment of what is being asserted. For instance this line should be changed to something like this.
```js
assert(!chest.slots[0], 'Slot 0 should be empty')
assert(chest.selectedItem?.count === 64 && chest.selectedItem?.type === 1, 'Selected item in chest should be stone (??) and have count 64')
```

Comment (IceTank) [inline test/test.js]: Please use the stoneId variable

File: test/test.js [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+  describe('mouseButton 5', () => {
+    it('number click from slot with item to slot with same item', () => {
+      chest.acceptClick(createClick(2, 5, 0))
+      assert(!chest.slots[0] && chest.slots[59]?.count === 4)
+    })
+
+    it('number click from slot with item to slot with different item', () => {
+      chest.acceptClick(createClick(2, 5, 5))
+      assert(!chest.slots[5] && chest.slots[59].type === 2 && chest.slots[59]?.count === 3 && chest.slots[54]?.count === 4 && chest.slots[54].type === 1)
```
Comment (IceTank) [inline test/test.js]:
```suggestion
      assert(!chest.slots[5] && chest.slots[59]?.type === 2 && chest.slots[59]?.count === 3 && chest.slots[54]?.count === 4 && chest.slots[54].type === 1)
```

File: test/test.js [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+describe('mode 3 | middle click', () => {
+  describe('mouseButton 2', () => {
+    it('get stack into selectedItem (gamemode 0)', () => {
+      inv.acceptClick(createClick(3, 2, inv.inventoryEnd - 2, true), 0)
+      assert(!inv.selectedItem)
+    })
+
+    it('get stack into selectedItem (gamemode 1)', () => {
+      inv.acceptClick(createClick(3, 2, inv.inventoryEnd - 2, true), 1)
+      assert(Item.equal(inv.selectedItem, new Item(1, inv.slots[inv.inventoryEnd - 2].stackSize)))
```
Comment (IceTank) [inline test/test.js]:
```suggestion
      assert(Item.equal(inv.selectedItem, new Item(1, inv.slots[inv.inventoryEnd - 2]?.stackSize)))
```

**Review by IceTank — CHANGES_REQUESTED** [review]

**Comment by IceTank** [comment]
Tests work for version 1.8 but every other version does not work. Have not tested on a real server.

File: test/test.js (hunk 16 above)
Comment (kaduvert) [reply test/test.js]: oops, didn't notice that id's are also used elsewhere.

will fix

**Comment by rom1504** [comment]
hey @kaduvert would you like to finish this ?

supporting more clicks seems like a great add

**Comment by kaduvert** [comment]
sorry for the delay guys.
this should fix everything bad with the old tests in terms of readability and it eliminates dependency on previously executed tests.

**Comment by kaduvert** [comment]
i now also added checks for updated but unasserted slots, meaning that if you test a click and it updates slots, you must assert all the updatedSlots or it fails.
this should help detect odd / unexpected behaviour.

i also fixed the version issue, tests now work with versions above 1.8.

File: index.js (hunk 8 above)
Comment (kaduvert) [reply index.js:3]: fixed.

File: lib/Window.js (hunk 5 above)
Comment (IceTank) [reply lib/Window.js:50]: Do you want to add offhand swap? It does not have to be in this pr but supporting it at some point would be nice.

File: index.d.ts [hunk 19]
```diff
@@ -49,24 +49,118 @@ export class Window<T = unknown> extends (EventEmitter as new <T>() => TypedEmit
...
+    /**
+     * accepts Clicks of with any mode, mouseButton and slot
+     * @param click click object to accept
+     * @param gamemode to know when certain clicks are allowed
+     */
+    acceptClick(click: Click, gamemode: number): void;
 
-    acceptClick(click: Click): void;
+    /**
+     * deprecated, call acceptClick(click) instead
```
Comment (IceTank) [inline index.d.ts]: A bit of a nit pick: There is a `@deprecated` jsdoc tag that strikes through a function when used in an ide. https://jsdoc.app/tags-deprecated.html. For this to work correctly the tag has to be in the index.d.ts file as this is where the ide's take there IntelliSense from and it has to be annotate for every function. A compact inline way to achieve this is by writing it like this
```js
/** @deprecated use {@link acceptClick} instead */
acceptNonInventorySwapAreaClick
```
the `@link` makes the linked function clickable in the IntelliSense preview.

Comment (IceTank) [reply index.d.ts]: The rest looks good to me. I have not tested it on an actual server and I won't be able to for some time. But if tests pass I think we can go ahead and merge it. We will just have to see if there are issues when people start to complain.

File: lib/Window.js (hunk 5 above)
Comment (IceTank) [reply lib/Window.js:50]: I don't consider this as a blocker

File: lib/Window.js [hunk 20]
```diff
@@ -22,106 +22,219 @@ module.exports = (Item) => {
...
+          this.mouseClick(click)
+          break
+
+        case 1:
+          assert.ok(mouseButton <= 1, 'invalid operation')
+          this.shiftClick(click)
+          break
+
+        case 2:
+          assert.ok(mouseButton <= 8, 'invalid operation')
```
Comment (IceTank) [inline lib/Window.js:50]: It looks like mouseButton for above 8 is possible in at least 1.19.4 (Although not vanilla) https://web.archive.org/web/20230519040131/https://wiki.vg/Protocol#Click_Container (web archive because I can't seam to access wiki.vg). So some sort of switch for newer versions would be nice at some point.

**Comment by kaduvert** [comment]
so i see that mouseButtons higher than 8 are a thing in newer versions and even 40 for offhand, but i have no clue and i can not confidently implement it because i have almost no experience with minecraft above 1.8.
also, one has to keep in mind, that the offhand slot is picky with what it accepts.

**Comment by kaduvert** [comment]
let's not forget that the click modes implemented here already come with huge time-benefits in window management.

all this other stuff like mode 5 clicks or mouseButtons higher than 8 in mode 2 are definitely useful, but i think that you're getting to the point of diminishing returns implementing them. also, i personally neither use or need the unimplemented stuff currently.

the code is structured so that you can add new modes, like 5 and 6 without much friction, so anyone who wants to can do it easily.

**Comment by kaduvert** [comment]
prismarine-windows as a whole has not implemented any logic for handling special slots, like armor slots in your inventory or the bottom slots in the brewing stand, which only accept some items.

this is not a huge problem with click modes 0 and 2-6.
you **could** in theory desync your window from the server one by moving the wrong items into slots which don't accept it, but you'd have to explicitly trigger this behaviour by specifying the slot as some kind of target. so it's unlikely to happen to anyone by accident.

with mode 1 on the other hand, this behaviour **can** happen without intent, for example if you shift around a piece of armor in an inventory, on the server it will go into the respective armor slot (if it's free), but because the client does not handle these special slots, it will (if the slot was free) lead to desync.

**Comment by kaduvert** [comment]
so what i suggest for now is labeling shift clicks as unstable in inventorys that have special slots until proper handling is implemented.
I could implement this handling, but i don't know where to find the data for which slots accepts which types of item, or if this data even exists. If it doesn't, i also wouldn't know how to reliably produce it, especially because this potentially changes with every minecraft version.

other than that however, there is no issues i know of.

**Comment by IceTank** [comment]
> also, one has to keep in mind, that the offhand slot is picky with what it accepts

This is only true for bedrock. In Java the offhand slot is like any other inventory slot and accepts all items.

**Comment by IceTank** [comment]
If the user tries to shit click an item into a slot he is not suppose to I would consider it user error. Also there is a good chance the server just rejects the click resulting in no change in the inventory.

**Comment by kaduvert** [comment]
Whatever, but i think it should be noted somewhere in the docs, just so the user knows.
Also, if the server rejects, the client immediately exits with an error https://github.com/PrismarineJS/mineflayer/blob/f50d904fca965ad9aef06ce31b7acfa1a7e8baa9/lib/plugins/inventory.js#LL549C1-L555C8

As to offhand, i just don't know.
someone else probably knows it better, so i will not try to implement it.
If you really want to, i can try, but i'd rather not.

**Comment by IceTank** [comment]
> Also, if the server rejects, the client immediately exits with an error

This is not good. Those rejections happen a lot and should be catchable. 

> https://github.com/PrismarineJS/mineflayer/blob/f50d904fca965ad9aef06ce31b7acfa1a7e8baa9/lib/plugins/inventory.js#LL549C1-L555C8

This peace of code is truly awful and should be refactored. But this is a concern for mineflayer not prismarine windows.

The offhand does accept all items. You can test it with a vanilla client in single player. But implementing that feature is not a priority I would say. We can add it at some later date.

**Comment by kaduvert** [comment]
Yeah, you're right with the offhand, i just checked.

Anyways, are there any more issues or can this be merged?

**Comment by rom1504** [comment]
Looks pretty good to me.

I think maybe one thing to try is open a PR in mineflayer using this to make sure it breaks no test.

Besides that I think I'll merge in next days if nobody has something to say

**Comment by kaduvert** [comment]
did it: https://github.com/PrismarineJS/mineflayer/pull/3102

all tests green @rom1504

**Review by rom1504 — APPROVED** [review]

**Comment by rom1504** [comment]
thanks for all the work on this
